benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,146 @@
+            benchmark: The benchmark definition that produced the result.
+        """
+
+
+@define
+class S3Persistence(PersistenceInterface):
+    """Class for persisting experiment results in an S3 bucket."""
+
+    _bucket_name: str = field(validator=instance_of(str), init=False)
+    """The name of the S3 bucket where the results are stored."""
+
+    _object_session = boto3.session.Session()
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    _branch: str = field(validator=instance_of(str), init=False)
+    """The branch of the BayBE library from which the workflow was started."""
+
+    _workflow_run_identifier: str = field(validator=instance_of(str), init=False)
+    """The identifier of the workflow run."""
+
+    @_bucket_name.default
+    def _default_bucket_name(self) -> str:
+        ENVIRONMENT_NOT_SET = "BAYBE_BENCHMARK_PERSISTENCE_PATH" not in os.environ
+        if ENVIRONMENT_NOT_SET:
+            raise ValueError(
+                "The environment variable "
+                "BAYBE_BENCHMARK_PERSISTENCE_PATH is not set. "
+                "A bucket name must be provided."
+            )
+        return os.environ["BAYBE_BENCHMARK_PERSISTENCE_PATH"]
+
+    @_branch.default
+    def _default_branch(self) -> str:
+        repo = Repo(search_parent_directories=True)
+        current_branch = repo.active_branch.name
+        S3_PATH_FORBIDDEN_CHARACTERS = ["/", "\\"]
+        sanitized_branch = ""
+        for char in S3_PATH_FORBIDDEN_CHARACTERS:
+            sanitized_branch = current_branch.replace(char, "-")
+        return sanitized_branch
+
+    @_workflow_run_identifier.default
+    def _default_workflow_id(self) -> str:
+        if "GITHUB_RUN_ID" not in os.environ:
+            raise ValueError("The environment variable GITHUB_RUN_ID is not set.")
+        return os.environ["GITHUB_RUN_ID"]
+
+    @override
+    def persist_new_result(self, result: Result, benchmark: Benchmark) -> None:
+        """Store the result of a benchmark.
+
+        This method will store the result of a benchmark in an S3 bucket.
+        The key of the object will be the experiment identifier, the branch, the
+        BayBE-version, the start datetime, the commit hash and the workflow
+        run identifier.
+
+        Args:
+            result: The result to be persisted.
+            benchmark: The benchmark definition that produced the result.
+        """
+        client = self._object_session.client("s3")
+
+        experiment_identifier = result.benchmark_identifier
+        metadata = result.metadata
+
+        bucket_path_key = (
+            f"{experiment_identifier}/{self._branch}/{metadata.latest_baybe_tag}/"
+            + f"{metadata.start_datetime.isoformat()}/"
+            + f"{metadata.commit_hash}/{self._workflow_run_identifier}"
+        )
+
+        result_dict = result.to_dict() | benchmark.to_dict()
```

Comment:
Technically, this makes the name of the function invalid, as you do not only store the `result` but also the `benchmark` information. After having a look at the class structures for those, I guess that this is currently necessary as some of the information that seems to be relevant for some of the parts yet to come are part of the `Benchmark` instead of the `Result`. So this makes me think if we should either (i) store such relevant information also in the `ResultMetadata` or (ii) rename the current function/workflow to something like `persist_result_and_benchmark`. However, let's discuss these options and further alternatives first here before implementing any changes (also summoning @AdrianSosic for this discussion as this also touches the topic of "which class should be responsible for what"). Happy to hear opinions :)

Reply:
This comment is by now outdated since there is a discussion regarding this at another place of the new code.